src/__experimental__/components/select-async/select-async.spec.js

```diff
@@ -0,0 +1,170 @@
+    expect(render(undefined, shallow)).toMatchSnapshot();
+  });
+
+  describe('on open', () => {
+    it('fetches data for the first page of results', async () => {
+      mockResponse();
+      const wrapper = await openResults(render());
+      expect(axios.get).toHaveBeenCalledWith(
+        endpoint, validRequest({ page: 1 })
+      );
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+  });
+
+  describe('no results', () => {
+    it('it renders with null', async () => {
+      mockResponse(validResponse([]));
+      const wrapper = await openResults(render());
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+  });
+
+  describe('custom children render', () => {
+    it('uses a custom method to render the children', async () => {
+      mockResponse();
+      const props = { children: items => items.map(() => <Option>custom!</Option>) };
+      const wrapper = await openResults(render(props));
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+  });
+
+  describe('custom request', () => {
+    it('allows developers to customise the request options', async () => {
+      mockResponse();
+      const props = { onRequest: () => ({ customRequest: true }) };
+      await openResults(render(props));
+      expect(axios.get).toHaveBeenCalledWith(
+        endpoint, { customRequest: true }
+      );
+    });
+  });
+
+  describe('custom response', () => {
+    it('throws an error if the response is not supported', async () => {
+      mockResponse({ data: {} });
+      try {
+        await render().instance().fetchData({ page: 1 });
+      } catch ({ message }) {
+        expect(message).toEqual(`The server response does not contain an $items attribute.
+
+The SelectAsync component expects the data to be in the format of:
+
+  $items: [Array] - required, the items to render
+  $page: [Integer] - optional, for paginated responses
+  $total: [Integer] - optional, for paginated responses
+
+If your API response does not match this, you can modify it using the 'onResponse' prop:
+
+<SelectAsync onResponse={ response => ({ $items: response.myItems }) } />`);
+      }
+    });
+
+    it('allows developers to customise the response format', async () => {
+      mockResponse();
+      const props = { onResponse: () => ({ $items: [{ displayed_as: 'custom!', id: '9' }] }) };
+      const wrapper = await openResults(render(props));
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+  });
+
+  describe('pagination', () => {
+    it('adds the next page results to the list', async () => {
+      mockResponse();
+      const wrapper = await openResults(render());
+      await getNextPage(wrapper);
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+
+    it('does not refetch if a fetch is already in progress', async () => {
+      mockResponse();
+      const wrapper = await openResults(render());
+      expect(axios.get.mock.calls.length).toEqual(1);
+      getNextPage(wrapper);
```

Comment:
Missing an `await` here I think.

Reply:
nah this is the subtle difference between this test and the next, this one is testing if a fetch is already in progress (so i do not want to wait for it to finish) maybe i should write a comment next to this line to that effect

Reply:
✅